rust/kernel/user_ptr.rs

```diff
@@ -190,25 +237,35 @@ impl UserSlicePtrWriter {
     /// Returns `EFAULT` if the byte slice is bigger than the remaining size
     /// of the user slice or if the address does not currently point to mapped,
     /// writable memory.
-    pub fn write(&mut self, data: &[u8]) -> error::KernelResult<()> {
-        if data.len() > self.1 || data.len() > u32::MAX as usize {
+    pub fn write_slice(&mut self, data: &[u8]) -> KernelResult<()> {
+        // SAFETY: The input buffer is valid as it's coming from a live reference.
+        unsafe { self.write_raw(data.as_ptr(), data.len()) }
+    }
+
+    /// Writes raw data to the user slice from a raw kernel buffer.
+    ///
+    /// # Safety
+    ///
+    /// The input buffer must be valid.
+    unsafe fn write_raw(&mut self, data: *const u8, len: usize) -> KernelResult<()> {
+        if len > self.1 || len > u32::MAX as usize {
             return Err(error::Error::EFAULT);
         }
-        let res = unsafe {
-            rust_helper_copy_to_user(
-                self.0,
-                data.as_ptr() as *const c_types::c_void,
-                data.len() as _,
-            )
-        };
+        let res = rust_helper_copy_to_user(self.0, data as _, len as _);
         if res != 0 {
             return Err(error::Error::EFAULT);
         }
         // Since this is not a pointer to a valid object in our program,
         // we cannot use `add`, which has C-style rules for defined
         // behavior.
-        self.0 = self.0.wrapping_add(data.len());
-        self.1 -= data.len();
+        self.0 = self.0.wrapping_add(len);
+        self.1 -= len;
         Ok(())
     }
+
+    /// Writes the contents of a plain old data (POD) type into the user slice.
+    pub fn write<T: Copy>(&mut self, data: &T) -> KernelResult<()> {
```

Comment:
Should this be `ReadableFromBytes` as well? `Copy` would let you write a struct with padding bytes that might be uninitialized!

Reply:
Paddings bytes are indeed problematic. I don't think `ReadableFromBytes` is sufficient to address this, as a struct containing `ReadableFromBytes` would be considered `ReadableFromBytes` but would still potentially leak data.

It happens to not be a problem in `binder` because it has explicit padding bytes (example [here](https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/android/binder.h#L101)).

I'll remove this function for now, I'll bring it back once I have a generic solution.